purchases/src/main/kotlin/com/revenuecat/purchases/Backend.kt

```diff
@@ -280,34 +296,67 @@ internal class Backend(
         attributes: Map<String, SubscriberAttribute>,
         appUserID: String,
         onSuccessHandler: () -> Unit,
-        onErrorHandler: (PurchasesError, syncedSuccessfully: Boolean) -> Unit
+        onErrorHandler: (PurchasesError, didBackendGetAttributes: Boolean, attributeErrors: List<SubscriberAttributeError>) -> Unit
     ) {
         enqueue(object : Dispatcher.AsyncCall() {
             override fun call(): HTTPClient.Result {
-                val attributesMap = attributes.map { (key, subscriberAttribute) ->
-                    key to subscriberAttribute.toBackendMap()
-                }.toMap()
                 return httpClient.performRequest(
                     "/subscribers/" + encode(appUserID) + "/attributes",
-                    mapOf("attributes" to attributesMap),
+                    mapOf("attributes" to attributes.toBackendMap()),
                     authenticationHeaders
                 )
             }
 
             override fun onError(error: PurchasesError) {
-                onErrorHandler(error, false)
+                onErrorHandler(error, false, emptyList())
             }
 
             override fun onCompletion(result: HTTPClient.Result) {
                 if (result.isSuccessful()) {
                     onSuccessHandler()
                 } else {
-                    onErrorHandler(result.toPurchasesError(), result.responseCode < HTTP_SERVER_ERROR_CODE)
+                    val error = result.toPurchasesError()
+                    var attributeErrors: List<SubscriberAttributeError> = emptyList()
+                    result.body?.takeIf { error.code == PurchasesErrorCode.InvalidSubscriberAttributesError }
+                        ?.let { body ->
+                            attributeErrors = body.getAttributeErrors()
+                        }
+                    onErrorHandler(error, result.responseCode < HTTP_SERVER_ERROR_CODE, attributeErrors)
                 }
             }
         })
     }
 
+    // Returns list of attribute errors found in the JSON object. The errors can be under
+    // ATTRIBUTES_ERROR_RESPONSE_KEY (for post receipt calls) or under
+    // ATTRIBUTE_ERRORS_KEY (for post subscriber attributes calls). If no attribute errors,
+    // returns an empty list
+    private fun JSONObject.getAttributeErrors(): List<SubscriberAttributeError> {
+        val attributeErrorsJSONObject =
+            if (has(ATTRIBUTES_ERROR_RESPONSE_KEY))
+                getJSONObject(ATTRIBUTES_ERROR_RESPONSE_KEY)
+            else
+                this
+
+        return if (attributeErrorsJSONObject.has(ATTRIBUTE_ERRORS_KEY)) {
```

Comment:
maybe reversing this if helps reduce nesting: 
```kotlin
if !attributeErrorsJSONObject.has(ATTRIBUTE_ERRORS_KEY) {
    return emptyList()
}

attributeErrorsJSONObject.getJSONArray(ATTRIBUTE_ERRORS_KEY)
    .let ...
```

 or better yet: 
```kotlin
val attributesJSONArray = attributeErrorsJSONObject.optJSONArray(ATTRIBUTE_ERRORS_KEY) ?: return emptyList()

attributesJSONArray
    .let ...

```

Reply:
Nice! I always forget about `opt...()`